qutip/qip/circuit.py

```diff
@@ -385,35 +546,47 @@ def add_circuit(self, qc, start=0):
-            elif gate.name in ["TOFFOLI"]:
-                self.add_gate(gate.name, gate.targets[0] + start,
-                              [gate.controls[0] + start,
-                               gate.controls[1] + start], None, None)
-            elif gate.name in ["FREDKIN"]:
-                self.add_gate(gate.name,
-                              [gate.targets[0] + start,
-                               gate.targets[1] + start],
-                              gate.controls + start, None, None)
-            elif gate.name in self.user_gates:
-                self.add_gate(
-                    gate.name, targets=gate.targets,
-                    arg_value=gate.arg_value)
+        for circuit_op in qc.gates_and_measurements:
+
+            if isinstance(circuit_op, Gate):
+                gate = circuit_op
+
+                if gate.name in ["RX", "RY", "RZ",
+                                "SNOT", "SQRTNOT", "PHASEGATE"]:
+                    self.add_gate(gate.name, gate.targets[0] + start, None,
+                                  gate.arg_value, gate.arg_label)
+                elif gate.name in ["X", "Y", "Z", "S", "T"]:
+                    self.add_gate(gate.name, gate.targets[0] + start, None,
+                                  None, gate.arg_label)
+                elif gate.name in ["CPHASE", "CNOT", "CSIGN", "CRX", "CRY",
+                                   "CRZ", "CY", "CZ", "CS", "CT"]:
+                    self.add_gate(gate.name, gate.targets[0] + start,
+                                  gate.controls[0] + start, gate.arg_value,
+                                  gate.arg_label)
+                elif gate.name in ["BERKELEY", "SWAPalpha", "SWAP", "ISWAP",
+                                   "SQRTSWAP", "SQRTISWAP"]:
+                    self.add_gate(gate.name,
+                                  [gate.targets[0] + start,
+                                   gate.targets[1] + start])
+                elif gate.name in ["TOFFOLI"]:
+                    self.add_gate(gate.name, gate.targets[0] + start,
+                                  [gate.controls[0] + start,
+                                   gate.controls[1] + start], None, None)
+                elif gate.name in ["FREDKIN"]:
+                    self.add_gate(gate.name,
+                                  [gate.targets[0] + start,
+                                   gate.targets[1] + start],
+                                  gate.controls + start, None, None)
+                elif gate.name in self.user_gates:
+                    self.add_gate(
+                        gate.name, targets=gate.targets,
+                        arg_value=gate.arg_value)
+            else:
+                measurement = circuit_op
+                self.add_measurement(
+                                measurement.name,
+                                targets=[measurement.targets[0] + start],
+                                classical_store=measurement.classical_store)
```

Comment:
I am a bit confused. How will we incorporate the shift in targets and controls if we are doing that !

Reply:
> I am a bit confused. How will we incorporate the shift in targets and controls if we are doing that !

We don't need to characterize the gate before copy it. All we need is to change the values of targets and controls. It's very simple if the default values are set to `[]`:
```
self.add_gate(gate.name, [target + start for target in gate.targets],
        [control + start for control in gate.controls], gate.arg_value,
        gate.arg_label)
```
By the way, I'm also considering provide more flexibilities in `add_circuit`. We can allow permutation of qubits instead of simply shift them, as in [my module](https://github.com/Canoming/VariationalQCircuit/blob/6447a8fbae09b30e469fb5072ae5dbfec50b2a44/variational_circuit/vcirc/base.py#L224).

Reply:
Or we copy it first and then change.
```python
new_gate = deepcopy(gate)
new_gate.targets = [t+start for t in new_gate.targets]
new_gate.controls= [t+start for t in new_gate.controls]
```
For `Measurement`, we don't have `controls` though, so only targets.

Reply:
It's also possible. But still, even for gates, if the default is `None`, we also need to check whether `[t+start for t in ...]` is valid. I'll suggest to change the default of targets and controls to empty lists. It looses nothing to store `targets` and `controls` as empty lists, even we don't need them. It works fine with measurements.

The only issue is that, it may require lots of dirty works. We have to change those `if gate.controls is None` everywhere. And there could be many (?) guys using this kind of codes in their projects.

Reply:
Just be careful with setting default _arguments_ to mutable types like "empty list".  It's fine for set parameters to be things like that, but they should _always_ be constructed fresh for each instance, otherwise it's quite easy to introduce a problem like this:
```ipython
In [1]: class A:
   ...:     def __init__(self, x=[]):
   ...:         self.x = x
   ...: a = A()
   ...: b = A()
   ...: a.x.append(1)
   ...: a.x, b.x
Out[1]: ([1], [1])
```

Much better to do:
```ipython
In [2]: class B:
   ...:     def __init__(self, x=None):
   ...:         self.x = x if x is not None else []
   ...: a = B()
   ...: b = B()
   ...: a.x.append(1)
   ...: a.x, b.x
Out[2]: ([1], [])
```
or something like this.

By the way, an alternative for the copying could be to give `Gate` and `Measurement` a `copy()` method and a `shift(n_qubits)` method, where each instance then can have its own logic for how this is implemented (e.g. `[el.copy().shift(n) for el in circuit_elements]`).

I'm glad I saw this conversation, though, because I'll absolutely have to check that the new data structures can support deep copy operations - being low-level C extension types, we  store non-reference-counted pointers to memory for speed, and that's anathema to black-box copy methods in a GC environment.

Reply:
Excellent catch @jakelishman, I was already making that error with the `cbits` parameter, and although it was probably not an immediate issue ! 

Re: changing the `add_circuit` function, given that it is slightly tangential to the purposes of this PR, maybe we should keep it for a subsequent PR or even package it with the cleanup PR @BoxiLi  @Canoming, what do you think ?

Reply:
@jakelishman Thanks for the comment. The code is still simple even if the default is `None`. 
```python
if new gate.targets is not None:
    new_gate.targets = [t+start for t in new_gate.targets]
```
So, I guess we can just leave the arguments like that.

@sarsid @BoxiLi  I think it's a good idea to leave it to the cleanup PR. It has nothing to do with the performance or new functionality, unless someone is interested in writing the `copy()` functions. That sounds like a hard job.

Reply:
Oh yeah, you don't need to worry for stuff like that - it's fine for `new_gate.targets` to be `[]` and possibly even better, since it's then uniform - I was just going for caution when _setting_ that default in the function signature.

Reply:
Agreed with all of you, personally I think leaving the default to be `None` is a safer choice and indeed let's leave it to a cleanup PR.